pkg/dataobj/internal/sections/logs/table_merge.go

```diff
@@ -0,0 +1,163 @@
+package logs
+
+import (
+	"cmp"
+	"container/heap"
+	"context"
+	"fmt"
+
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/dataset"
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/metadata/logsmd"
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/result"
+)
+
+// mergeTables merges the provided sorted tables into a new single sorted table
+// using heap sort.
+func mergeTables(buf *tableBuffer, pageSize int, compressionOpts dataset.CompressionOptions, tables []*table) (*table, error) {
+	buf.Reset()
+
+	var (
+		streamIDBuilder  = buf.StreamID(pageSize)
+		timestampBuilder = buf.Timestamp(pageSize)
+		messageBuilder   = buf.Message(pageSize, compressionOpts)
+	)
+
+	var (
+		// columnSets holds the set of columns per table.
+		columnSets   = make([][]dataset.Column, 0, len(tables))
+		tablePullers = make([]pullRow, 0, len(tables)) // Pull iterator per table.
+	)
+	for _, t := range tables {
+		dsetColumns, err := result.Collect(t.ListColumns(context.Background()))
+		if err != nil {
+			return nil, err
+		}
+
+		seq := dataset.Iter(context.Background(), dsetColumns)
+		next, stop := result.Pull(seq)
+		defer stop()
+
+		columnSets = append(columnSets, dsetColumns)
+		tablePullers = append(tablePullers, next)
+	}
+
+	var h rowHeap
```

Comment:
if rows are sorted within their table I think you can use a loser tree instead so that you don't need a heap container see https://github.com/bboreham/go-loser

It needs the table to implement Sequence[Row] we use it here in Loki already https://github.com/grafana/loki/blob/main/pkg/iter/entry_iterator.go#L85-L86

The advantage is that you don't need the heap operation for each row !